iothub/service/src/Common/AsyncResult.cs

```diff
@@ -281,6 +281,7 @@ private bool TryContinueHelper(IAsyncResult result, out AsyncCompletion callback
             }
             else
 #endif
+            if (!result.CompletedSynchronously)
```

Comment:
if you expand the file to see the method `TryContinueHelper` you will notice that we will have an unreachable code for all frameworks other than NET451. 
This method used to look like this:

`        private bool TryContinueHelper(IAsyncResult result, out AsyncCompletion callback)
        {
            if (result == null)
            {
                throw Fx.Exception.AsError(new InvalidOperationException(CommonResources.InvalidNullAsyncResult));
            }

            callback = null;

            if (result.CompletedSynchronously)
            {
#if NET451
                // Once we pass the check, we know that we own forward progress, so transactionContext is correct. Verify its state.
                if (_transactionContext != null)
                {
                    if (_transactionContext.State != TransactionSignalState.Completed)
                    {
                        ThrowInvalidAsyncResult("Check/SyncContinue cannot be called from within the PrepareTransactionalCall using block.");
                    }
                    else if (_transactionContext.IsSignalled)
                    {
                        // This is most likely to happen when result.CompletedSynchronously registers differently here and in the callback, which
                        // is the fault of 'result'.
                        ThrowInvalidAsyncResult(result);
                    }
                }
#endif
            }
#if NET451
            else if (ReferenceEquals(result, _deferredTransactionalResult))
            {
                // The transactionContext may not be current if forward progress has been made via the callback. Instead,
                // use deferredTransactionalResult to see if we are supposed to execute a post-transaction callback.
                //
                // Once we pass the check, we know that we own forward progress, so transactionContext is correct. Verify its state.
                if (_transactionContext == null || !_transactionContext.IsSignalled)
                {
                    ThrowInvalidAsyncResult(result);
                }
                _deferredTransactionalResult = null;
            }
#endif
            else
            {
                return false;
            }

            callback = GetNextCompletion();
            if (callback == null)
            {
                ThrowInvalidAsyncResult("Only call Check/SyncContinue once per async operation (once per PrepareAsyncCompletion).");
            }
            return true;
        }`

Reply:
As you can see, the `if (result.CompletedSynchronously)` statement was outside of the net framework condition so the `else` after `#endif` made sense. 
but now it will always return false for all frameworks other than 451 regardless.

Reply:
Good point. Do we need to worry about the `else if` clause too?

Reply:
excellent question, I didn't think about that part. let me revert this code and re-do the whole method.

Reply:
Sorry for creating this problem in the first place.

Reply:
no worries at all !

Reply:
ok, with _transactionContext gone, I think I got parity with the original functionality.